torch/export/exported_program.py

```diff
@@ -118,6 +120,130 @@ def _fx_collection_equivalence_fn(
     return spec1_type is spec2_type and spec1_context == spec2_context
 
 
+def _skip_autograd_kernel(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    with torch._C._AutoDispatchBelowAutograd():
+        result = kernel(*args, **kwargs)
+        # TODO (tmanlaibaatar) Remove this as it is copied from torch/_higher_order_ops/utils.py
+        # Ideally it should re-use the custom op autograd error directly
+        flat_operands = pytree.arg_tree_leaves(*args)
+        if torch.is_grad_enabled() and any(
+            f.requires_grad for f in flat_operands if isinstance(f, torch.Tensor)
+        ):
+            err_fn = torch._C._functions.DelayedError(
+                f"Autograd not implemented for {str(kernel)}",
+                1,
+            )
+
+            def fake_requires_grad(tensor):
+                if torch.is_floating_point(tensor) or torch.is_complex(tensor):
+                    tensor = tensor.detach()
+                    tensor.requires_grad = True
+                return tensor
+
+            return pytree.tree_map_only(
+                torch.Tensor, lambda x: err_fn(fake_requires_grad(x)), result
+            )
+        return result
+
+
+def _register_cia_to_meta(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(
+            torch._C.DispatchKey.CompositeImplicitAutograd, *args, **kwargs
+        )
+    else:
+        return NotImplemented
+
+
+@contextmanager
+def override_composite_implicit_decomp(ops_to_preserve):
+    # This function overrides CompositeImplicitAutograd decomp for
+    # functional composite ops that user specified. Ideally we want to not-decompose
+    # ALL composite ops but today's C++ functinalization relies on
+    # the fact that it is working with the opset after decomp is run.
+    # Hence we can only do it for functional ops. One caveat is that
+    # there are some composite ops that lie about their schema (claimed to be
+    # functional but not really aka dropout), for these cases, we just decompose.
+    saved_tables = {}
+    patched_ops = set()
+    for op_overload in ops_to_preserve:
+        # Our strategy for deciding if we can preserve CIA is following:
+        # 1. The op should be known statically that it is functional
+        # 2. If it is maybe aliasing, we decompose because we must know if an op
+        #    is mutating or aliasing.
+        # TODO (tmanlaibaatar) make this utility function and share it with functional_tensor
```

Comment:
Could you file an issue for this TODO (and any others in this PR) and assign yourself so we can better track them?

Reply:
Done!